Sources/AutoLayout/Constrain.swift

```diff
@@ -41,6 +41,45 @@ public final class ConstraintGroup {
     }
 }
 
+public final class ConstraintGroupToggle<T: Hashable> {
+
+    private var constraintGroups: [T: ConstraintGroup] = [:]
+    private let makeConstraintGroup: (T) -> ConstraintGroup?
+
+    public init(initial: T? = nil, _ makeConstraintGroup: @escaping (T) -> ConstraintGroup?) {
+
+        self.makeConstraintGroup = makeConstraintGroup
+
+        if let initial = initial {
+            activate(initial)
+        }
+    }
+
+    public func activate(_ key: T) {
+
+        var previousActiveConstraint: ConstraintGroup?
+
+        constraintGroups.forEach {
+            
+            if $0 != key && $1.isActive {
+                $1.isActive = false
+                previousActiveConstraint = $1
+            }
+        }
+
+        if let constraintGroup = constraintGroups[key] {
+            constraintGroup.isActive = true
+        } else {
+            if let constraintGroup = makeConstraintGroup(key) {
+                constraintGroups[key] = constraintGroup
+                constraintGroup.isActive = true
+            } else {
+                previousActiveConstraint?.isActive = true
```

Comment:
As a user of this API if for some reason the `makeConstraintGroup` returns `nil` I would expect the behaviour of `activate` to still disable all the previous constraints, no? It just wouldn't apply any constraints, same as if we return `ConstraintGroup()` and same as if it's the initial `key`.

Maybe `constraintGroups` should be `[T: ConstraintGroup?]` to avoid re-running `makeConstraintGroup` when it returns `nil`.

Reply:
@IvoPintodaSilva @filipe-lemos I had tried both of these iterations:

Filipe's was my first iteration, I went with keeping the last constraint mainly because toggle made me feel that there should always be an active ConstraintGroup, that being said we can init without a ConstraintGroup active.   

In regards to Ivo's approach I also considered it and it might be the best way to lead the programmer away from code smells, like using Non Finite Key Sets.

I am leaning towards returning an error, but not sure if I should undo the previous constraints in that case.

Reply:
In my opinion, from a caller's perspective, if a method that I'm calling yields an error, I expect no changes in the internal state.

Reply:
But that's assuming a value of `nil` in a `makeConstraintGroup` closure is an error, I would consider that as "for this key there are no constraints"

Reply:
I have now migrated a few use cases in our app and the only situation where I returned nil was because I have to capture self `guard let self = self else { return nil }`

I considered having the closure itself throw `makeConstraintGroup: (T) throws -> ConstraintGroup?`, but felt a bit like overkill.

So decision time, I think we can go with Filipe's more general approach and iterate in the future if we feel the need. If there are now further objections 😄